### DISCUSSION TITLE
Release v0.3.0: Customizable Server-Side Prompts

### DISCUSSION THREAD

**Discussion by icereed** [icereed/paperless-gpt#16] [Announcements]
# 📢 Release Notes: paperless-gpt v0.3.0

We are excited to announce the release of **paperless-gpt v0.3.0**! This version brings significant enhancements, including customizable server-side prompts and improvements to tag handling. These updates provide greater flexibility and control over how you generate document titles and tags using AI.

### 1. **Customizable Server-Side Prompts**

- **User-Defined Prompt Templates**: You can now modify the prompt templates used for generating document titles and tags directly on the server side.

  - **Prompt Templates**: The application uses Go's `text/template` for the prompt templates.

  - **Persistent Storage**: Templates are stored in the `prompts` directory, allowing you to customize them by editing the files.

  - **Template Variables**: Use variables like `{{.Language}}`, `{{.Content}}`, `{{.AvailableTags}}`, and `{{.Title}}` within your templates to tailor the prompts.

  - **Advanced Functions**: Leverage functions from the [Sprig](http://masterminds.github.io/sprig/) library for more advanced templating capabilities.

- **Easy Integration**: Simply edit the template files to adjust how the AI generates suggestions, making the system more adaptable to your specific needs.

### 2. **Improved Tag Handling**

- **API Pagination Fix**: Resolved an issue where not all tags were fetched from paperless-ngx due to pagination limitations.

  - **Comprehensive Tag Fetching**: The application now correctly handles API pagination, ensuring that all available tags are retrieved and considered during tag suggestion generation.

- **Enhanced Accuracy**: With all tags available, the AI can suggest more accurate and relevant tags for your documents.

## 🛠 How to Update

To update to the latest version of paperless-gpt, please pull the latest Docker image:

```bash
docker pull icereed/paperless-gpt:latest
```

If you're using Docker Compose, update your `docker-compose.yml` file if necessary and redeploy the stack:

```bash
docker-compose pull
docker-compose up -d
```

### Mounting the Prompts Directory

To enable custom prompt templates, you need to mount a local `prompts` directory into the container.

**Docker Compose Example:**

```yaml
services:
  paperless-gpt:
    image: icereed/paperless-gpt:latest
    # ... (other configurations)
    volumes:
      - ./prompts:/app/prompts # Mount the prompts directory
```

**Docker Run Command Example:**

```bash
docker run -d \
  # ... (other configurations)
  -v $(pwd)/prompts:/app/prompts \
  paperless-gpt
```

### Editing the Prompt Templates

1. **Create the `prompts` Directory Locally:**

   ```bash
   mkdir prompts
   ```

2. **Start the Container:**

   When you first start the container with the prompts directory mounted, it will automatically create the default template files in your local `prompts` directory if they do not exist.

3. **Edit the Template Files:**

   - Open `prompts/title_prompt.tmpl` and `prompts/tag_prompt.tmpl` with your favorite text editor.
   - Modify the templates using Go's `text/template` syntax.
   - Save the changes.

4. **Restart the Container (if necessary):**

   The application automatically reloads the templates when it starts. If the container is already running, you may need to restart it to apply the changes.

#### Template Syntax and Variables

- **For `title_prompt.tmpl`:**

  - `{{.Language}}`: The language specified in `LLM_LANGUAGE` (default is `English`).
  - `{{.Content}}`: The content of the document.

- **For `tag_prompt.tmpl`:**

  - `{{.Language}}`: The language specified in `LLM_LANGUAGE`.
  - `{{.AvailableTags}}`: A comma-separated list of available tags from paperless-ngx.
  - `{{.Title}}`: The suggested title for the document.
  - `{{.Content}}`: The content of the document.

**Note:** Advanced users can utilize additional functions from the [Sprig](http://masterminds.github.io/sprig/) template library, as it is included in the application.

## 📚 Documentation

For detailed instructions on installation, configuration, and customizing prompt templates, please refer to the [README](https://github.com/icereed/paperless-gpt/blob/main/README.md) on our GitHub repository.

## 💡 Feedback and Contributions

We appreciate your feedback! If you encounter any issues or have suggestions for new features, please open an issue on our [GitHub Issues](https://github.com/icereed/paperless-gpt/issues) page.

Contributions are welcome. Please read our [contributing guidelines](https://github.com/icereed/paperless-gpt/blob/main/CONTRIBUTING.md) before submitting a pull request.

## 📦 What's Next

We are continuously working to improve paperless-gpt. Upcoming features and enhancements may include:

- **UI Support for Prompt Editing**: Providing an in-app interface to edit prompt templates without the need to access the file system.

- **Manual Dark Mode Toggle**: Allowing users to override the system's dark mode setting within the application.

- **Additional LLM Support**: Expanding compatibility with more Large Language Models.

- **Performance Improvements**: Optimizing the application for faster processing and better resource utilization.

---

Thank you for your continued support and for using paperless-gpt to enhance your paperless-ngx experience!

<hr /><em>This discussion was created from the release <a href='https://github.com/icereed/paperless-gpt/releases/tag/v0.3.0'>Customizable Server-Side Prompts</a>.</em>

**Comment by vibe-code-repo** [comment]
Great Job, thanks :)
